linera-core/src/chain_worker/state.rs

```diff
@@ -346,21 +343,17 @@ where
         }
 
         let pending_blobs = &self.chain.manager.get().pending_blobs;
-        let tasks = self
+        let blob_ids = self
             .recent_hashed_blobs
             .subtract_cached_items_from::<_, Vec<_>>(required_blob_ids, |id| id)
             .await
             .into_iter()
             .filter(|blob_id| !pending_blobs.contains_key(blob_id))
-            .map(|blob_id| {
-                self.storage
-                    .contains_blob(blob_id)
-                    .map(move |result| (blob_id, result))
-            });
-
+            .collect::<Vec<_>>();
+        let results = self.storage.contains_blobs(blob_ids.clone()).await?;
         let mut missing_blobs = vec![];
-        for (blob_id, result) in future::join_all(tasks).await {
-            if !result? {
+        for (blob_id, result) in blob_ids.into_iter().zip(results) {
```

Comment:
If we are `zip`ping like this in most (all?) call sites, would it make sense to return an actual map, or `Vec<BlobId, bool>`, instead? It would probably be less error-prone.